eduiddashboard/utils.py

```diff
@@ -191,3 +194,195 @@ def convert_to_localtime(dt):
+        return sanitize_input(request.POST[post_parameter],
+                              percent_encoded=use_percent_encoding)
+    except KeyError:
+        logger.warn('An unexpected error occurred: POST parameter {!r} '
+                    'could not be found in the request.'.format(post_parameter))
+
+        # Re-raise the exception to not change the
+        # expected logic of the wrapped function.
+        raise
+
+    except UnicodeDecodeError:
+        logger.warn('A malicious user tried to crash the application '
+                    'by sending non-unicode input in a POST request')
+        raise HTTPBadRequest("Non-unicode input, please try again.")
+
+
+def sanitize_input(untrusted_text, strip_characters=False, percent_encoded=False):
+    """
+    Sanitize user input by escaping or removing potentially
+    harmful input using a whitelist-based approach with
+    bleach as recommended by OWASP.
+
+    :param untrusted_text User input to sanitize
+    :param strip_characters Set to True to remove instead of escaping harmful input
+    :param percent_encoded Set to False if input is not percent encoded
+    :return: Sanitized user input
+
+    :type untrusted_text: str | unicode
+    :rtype: str | unicode
+    """
+
+    if untrusted_text is None:
+        # If we are given None then there's nothing to clean
+        return None
+
+    elif percent_encoded:
+        # If the untrusted_text is percent encoded we have to:
+        # 1. Decode it so we can process it.
+        # 2. Encode it to UTF-8 since bleach assumes this encoding
+        # 3. Clean it to remove dangerous characters.
+        # 4. Percent encode it before returning it back.
+
+        decoded_text = unquote(untrusted_text)
+
+        if not isinstance(decoded_text, unicode):
+            decoded_text_in_utf8 = decoded_text.encode("UTF-8")
+        else:
+            decoded_text_in_utf8 = decoded_text
+
+        cleaned_text = __safe_clean(decoded_text_in_utf8, strip_characters)
+        percent_encoded_text = quote(cleaned_text)
+
+        if decoded_text_in_utf8 != cleaned_text:
+            logger.warn('Some potential harmful characters were '
+                        'removed from untrusted user input.')
+
+        return percent_encoded_text
+
+    else:
+        # If the untrusted_text is not percent encoded we only have to:
+        # 1. Encode it to UTF-8 since bleach assumes this encoding
+        # 2. Clean it to remove dangerous characters.
+
+        if not isinstance(untrusted_text, unicode):
+            text_in_utf8 = untrusted_text.encode("UTF-8")
+        else:
+            text_in_utf8 = untrusted_text
+
+        cleaned_text = __safe_clean(text_in_utf8, strip_characters)
+
+        if text_in_utf8 != cleaned_text:
+            logger.warn('Some potential harmful characters were '
+                        'removed from untrusted user input.')
+
+        return cleaned_text
+
+
+def __safe_clean(untrusted_text, strip_characters=False):
```

Comment:
I read PEP-8 as suggesting one leading underscore for a private function name, not two.

"_single_leading_underscore : weak "internal use" indicator. E.g. from M import * does not import objects whose name starts with an underscore."